articles/200_coexistence_and_migration.md

```diff
@@ -0,0 +1,247 @@
+
+**Decision:** Use separate namespaces for packages in ROS 1 and ROS 2 (this is currently already the case).
+
+### Namespace of rosdep keys
+
+The next question is very similar and is about the namespace of the `rosdep` keys.
+Should ROS 1 and ROS 2 use a different set of `rosdep` rules or should they share the same database?
+
+#### Same key vs. different key
+
+If a ROS package declares a dependency on the `rosdep` key `python-foo` that maps on e.g. Ubuntu Xenial to the Python 2 package `foo`.
+In order to work for ROS 2 it needs to depend on the Python 3 version of the package instead.
+
+Currently the mapping in `rosdep` only depends on the platform and doesn't consider which ROS distribution is being used.
+
+If the migration to ROS 2 requires the key to be changed to `python3-foo` that implies that any package which depends on Python packages does need a modification in the manifest.
+
+The same conceptional problems already exist in ROS 1.
+When being built on platforms which use Python 3 they already map the keys to Python 3 packages.
+If this problem would be address within `rosdep` to consider the Python version as part of the platform that would allow Python 2 as well as Python 3 packages to be resolved using the same `rosdep` key.
+This would also allow building / using / testing ROS 1 with Python 3.
+
+Another example is the mapping of `rosdep` key to the Qt 4 or Qt 5 version of a dependency.
+Package try to avoid changing the rosdep key used for their dependencies but rosdep only considers the platform without the context of the ROS distribution.
+
+**Proposal:** Change request for `rosdep` to consider information beyond the platform to determine the mapping.
+
+**Decision:** With that change the `rosdep` rules should share a single namespace (this is currently already the case).
+
+### Package manifest
+
+In ROS 1 two versions of the package manifest are supported: format 1 specified in [REP 127](http://www.ros.org/reps/rep-0127.html) and format 2 specified in [REP 140](http://www.ros.org/reps/rep-0140.html).
+While ROS 2 only supports format 2 there is no technical limitation behind that design decision.
+It simply reduces the effort to distinguish the different semantics of some depend tags in various parts of the software.
+
+It might be the case that a to-be-specified format 3 will be created to add support for the desired features to describe [dependency groups]() as well as declaring an [ABI version]().
+
+#### ROS dependencies
+
+While a ROS 1 package might depend on `roscpp` and `message_generation` a ROS 2 package depends on `rclcpp` and `rosidl_message_generation` instead.
+
+#### Build type
+
+Almost all build types for ROS 1 and ROS 2 are disjoint except the build type `cmake` which can be processed by both build tools.
+Since it is not desired to change existing `catkin` / `ament_cmake` packages to use the build type `cmake` those are currently different for each ROS version.
```

Comment:
Those are listed in this article:

* the rosdep keys for the non-ROS dependencies
* ROS dependencies like `roscpp` and `message_generation`
* message definitions must only use the ROS 1 feature set
* the include paths for message headers
* the includes and types of API which is different in ROS 1 and ROS 2
* the code base using Boost vs. C++11 types

How is `COMMAND catkin_package` going to work? What happens before? How does a package know if it should build for ROS 1 or ROS 2?

Reply:
This isn't something I think should be done, I'm just trying to understand why it can't be done. Feel free to ignore this.

I was assuming the developer would have either sourced the ros2 workspace and used ament, or sourced a ros1 workspace and used catkin.

On those bullets It seems like the package.xml dependencies being different for ROS1 and ROS2 is really the only thing that can't be worked around. C++ API incompatibilities can be preprocessor macros and `configure_file`. Message definitions being restricted is inconvenient, but since ROS1 .msg files can be anywhere a package could have a ros1 message directory and use `msg/` for ros2 messages